### PR TITLE
Update vecgeom to 1.1.20

### LINKED ISSUE #7734: Migration VecGeom to the new version 1.1.20
New version of VecGeom https://gitlab.cern.ch/VecGeom/VecGeom 1.1.20 is available. It includes fixes, the most important fix of CutTubs for stack tracks [VECGEOM-540]. It is needed to migrate G4VECGEOM branch to this VecGeom version. After testing it would be possible to consider it for the master. Recent vecgeom.spec modification in merged #7730 is not needed.

### PR DESCRIPTION
**Pull request by iarspider**
Closes #7734

### CODE REVIEW

**Comment by iarspider** [comment]
please test

File: vecgeom.spec [hunk 1]
```diff
@@ -24,7 +24,6 @@ cmake ../%{n}-%{realversion} \
   -DCMAKE_BUILD_TYPE=Release \
   -DNO_SPECIALIZATION=ON \
   -DBACKEND=Scalar \
-  -DGEANT4_USE_USOLIDS="CONS;POLYCONE;POLYHEDRA" \
```
Comment (smuzaffar) [inline vecgeom.spec:27]: @iarspider , if I understad the @civanch comment https://github.com/cms-sw/cmsdist/issues/7734 , then we only need to remove the changes of `vecgeom.spec`. So I think this change we still need ... right @civanch ? I might be wrong, so please wait for @civanch response

**Comment by smuzaffar** [comment]
@iarspider , looks like we need veccore?

File: geant4.spec [hunk 2]
```diff
@@ -49,7 +49,6 @@ cmake ../%{n}.%{realversion} \
   -DGEANT4_BUILD_TLS_MODEL:STRING="global-dynamic" \
   -DGEANT4_ENABLE_TESTING=OFF \
   -DGEANT4_BUILD_VERBOSE_CODE=OFF \
-  -DGEANT4_BUILD_BUILTIN_BACKTRACE=OFF \
```
Comment (civanch) [inline geant4.spec]: this line better do not remove - Geant4 native backtrace may mess up CMS exceptions.

Comment (smuzaffar) [inline geant4.spec]: @iarspider , if I understad the @civanch comment https://github.com/cms-sw/cmsdist/issues/7734 , then we only need to remove the changes of vecgeom.spec. So I think this change we still need ... right @civanch ? I might be wrong, so please wait for @civanch response

**Comment by civanch** [comment]
geant4.spec should not be modified: -DGEANT4_BUILD_BUILTIN_BACKTRACE=OFF is useful, because it means all exceptions will be handled by CMSSW, no extra backtrace from Geant4.

-DGEANT4_USE_USOLIDS="CONS;POLYCONE;POLYHEDRA"  better to remove in order do not mess up new version of VecGeom and this extra configuration. We may use it later but now we need to have a branch which can run big statistics smoothly and be minimally differ from vecgeom 1.1.18.

**Comment by iarspider** [comment]
> @iarspider , looks like we need veccore?

Yes, looks like so. I will look into it.

**Comment by civanch** [comment]
VecCore 0.8.0 is needed. Gabriele suggest following cmake options:

 -DBACKEND=YOUR-CHOSEN-BACKEND [Vc|Scalar]
 -DVECGEOM_VECTOR=YOUR-CHOSEN-VECTOR-INSTRUCTION-SET [native|sse2|avx2]
 -DVecCore_DIR=[YOUR--VECCORE-INSTALL-PATH]/lib/cmake/VecCore
 -DCMAKE_PREFIX_PATH=[YOUR-VC-INSTALL-PATH]
 -DDATA_DOWNLOAD=OFF
 -DGEANT4=OFF
 -DCMAKE_BUILD_TYPE=Release
 -DCMAKE_INSTALL_PREFIX=[YOUR-VECGEOM-INSTALL-PATH]

He said that the explicit path to VECCORE is needed.

**Comment by iarspider** [comment]
please test

**Comment by iarspider** [comment]
please test

**Comment by iarspider** [comment]
please test

**Comment by smuzaffar** [comment]
@iarspider , I think we need to add veccore toolfile too. It only provides header files ( i.e. no library) , so please add a toolfilewith only INCLUDE flag. Also update vecgeom tool file to use it `<use name="veccore"/>

**Comment by civanch** [comment]
should geant4 depends on veccore?

**Comment by iarspider** [comment]
please test

**Comment by smuzaffar** [comment]
please test

**Comment by smuzaffar** [comment]
please test

**Comment by smuzaffar** [comment]
@civanch , can you please check the relvals error here? these are failing/crashing in vecgeom

**Comment by civanch** [comment]
It seems that vecgeom::cxx::PolyhedronImplementation is responsible. I have informed G4 geometry experts but they have no idea - their tests using CMS geometry are running.

**Comment by civanch** [comment]
It seems that we do not have a line inside vecgeom.spec:

-DCMAKE_INSTALL_PREFIX=[YOUR-VECGEOM-INSTALL-PATH]

**Comment by smuzaffar** [comment]
@civanch , we have it here https://github.com/cms-sw/cmsdist/pull/7735/files#diff-a80b72200902a4302b1e1b82206de2248365671ed696edfd066a2f92470c5ce0R22 .

**Comment by civanch** [comment]
@smuzaffar , I am not an expert in cmake, does it use environment variables?

**Comment by smuzaffar** [comment]
@civanch , no we privide the full install path e.g.  [-DCMAKE_INSTALL_PREFIX=%{i}](https://github.com/cms-sw/cmsdist/blob/e4c2a0febab587658aef9b349f221bf1a5552784/vecgeom.spec#L22)  where `%{i}` is the full path where vecgeom should be installed.

**Comment by civanch** [comment]
@smuzaffar , I do not fully understand the build system but may be this line should be added to vecgeom.spec 

 -DVecCore_DIR=${VECCORE_ROOT}/lib/cmake/VecCore \

**Comment by smuzaffar** [comment]
this is exact cmake command we use
```
cmake ../vecgeom-v1.1.20 \
-DCMAKE_INSTALL_PREFIX=/data/cmsbld/jenkins/workspace/ib-run-pr-tests/testBuildDir/tmp/BUILDROOT/d6b0a3dec34119d2f295cd497ecdbd18/opt/cmssw/slc7_amd64_gcc10/external/vecgeom/v1.1.20-d6b0a3dec34119d2f295cd497ecdbd18 \
-DCMAKE_INSTALL_LIBDIR=/data/cmsbld/jenkins/workspace/ib-run-pr-tests/testBuildDir/tmp/BUILDROOT/d6b0a3dec34119d2f295cd497ecdbd18/opt/cmssw/slc7_amd64_gcc10/external/vecgeom/v1.1.20-d6b0a3dec34119d2f295cd497ecdbd18/lib \
-DROOT=OFF -DCMAKE_BUILD_TYPE=Release -DNO_SPECIALIZATION=ON \
-DBACKEND=Scalar -DVECGEOM_VECTOR=sse3 -DCMAKE_VERBOSE_MAKEFILE=TRUE \
-DCMAKE_CXX_STANDARD=17 -DGEANT4=OFF \
-DDATA_DOWNLOAD=OFF -DCMAKE_PREFIX_PATH=/data/cmsbld/jenkins/workspace/ib-run-pr-tests/testBuildDir/slc7_amd64_gcc10/external/veccore/0.8.0-5aaf04d3523c463c46dfff8c8fde7f5a
```

**Comment by smuzaffar** [comment]
I think it is not needed but sure we can try to add `-DVecCore_DIR=${VECCORE_ROOT}/lib/cmake/VecCore`

**Comment by smuzaffar** [comment]
please test

**Comment by civanch** [comment]
@smuzaffar , you was right - the recent modification change nothing. Gabriele saying that -DVECGEOM_VECTOR=sse3 may be a problem with Scalar backend.

@bsunanda trying install 1.1.20 privately and also fail. His installation instructions without VecCore. He is able to build and crash in a similar way at the first event.

**Comment by bsunanda** [comment]
This does not crash for testbeam application. I am not sure if it has something to do with overall CMS detector in polyhedron cases - but TB application also uses polyhedron - so I am confused
________________________________

**Comment by civanch** [comment]
@smuzaffar , @bsunanda , we discussed this Gabriele once more and he suggest to remove -DVECGEOM_VECTOR=sse3 , because it override Scalar option. He said that in the scalar mode this option should not present and from his point of view it is the only explanation.

**Comment by smuzaffar** [comment]
please test

**Comment by smuzaffar** [comment]
> @smuzaffar , @bsunanda , we discussed this Gabriele once more and he suggest to remove -DVECGEOM_VECTOR=sse3 , because it override Scalar option. He said that in the scalar mode this option should not present and from his point of view it is the only explanation.

ok thanks @civanch for the information, I have updated vecgeom to not set `VECGEOM_VECTOR` in `Scalar` mode.

**Comment by smuzaffar** [comment]
@civanch , removing ` -DVECGEOM_VECTOR=sse3 ` did not work . `cmake` command used for latest build is
```
cmake ../vecgeom-v1.1.20 \
  -DCMAKE_INSTALL_PREFIX=/data/cmsbld/jenkins/workspace/ib-run-pr-tests/testBuildDir/tmp/BUILDROOT/0f85be0e9f610c52060747075bed42c5/opt/cmssw/slc7_amd64_gcc10/external/vecgeom/v1.1.20-0f85be0e9f610c52060747075bed42c5 \
  -DCMAKE_INSTALL_LIBDIR=/data/cmsbld/jenkins/workspace/ib-run-pr-tests/testBuildDir/tmp/BUILDROOT/0f85be0e9f610c52060747075bed42c5/opt/cmssw/slc7_amd64_gcc10/external/vecgeom/v1.1.20-0f85be0e9f610c52060747075bed42c5/lib \
  -DROOT=OFF \
  -DCMAKE_BUILD_TYPE=Release \
  -DNO_SPECIALIZATION=ON \
  -DBACKEND=Scalar \
  -DCMAKE_VERBOSE_MAKEFILE=TRUE \
  -DCMAKE_CXX_STANDARD=17 \
  -DGEANT4=OFF \
  -DDATA_DOWNLOAD=OFF \
  -DVecCore_DIR=/data/cmsbld/jenkins/workspace/ib-run-pr-tests/testBuildDir/slc7_amd64_gcc10/external/veccore/0.8.0-5aaf04d3523c463c46dfff8c8fde7f5a/lib64/cmake/VecCore \
  -DCMAKE_PREFIX_PATH=/data/cmsbld/jenkins/workspace/ib-run-pr-tests/testBuildDir/slc7_amd64_gcc10/external/veccore/0.8.0-5aaf04d3523c463c46dfff8c8fde7f5a
```

**Comment by civanch** [comment]
@smuzaffar , Ben Morgan took 1.1.20 compile and run in ATHENA - he said it is working out of the box for them. It may be the case for previous versions of Geant4, not for Geant4 11.

We can do two possible checks:
1) make similar to this PR for vecgeom for the Geant4 master not the branch
2) I have the link to the fix of CutTubs, so only this fix can be applied. I would make PR on top of vecgeom 1.1.18 but I do not see vecgeom as a project within CMSSW. Does this exist?

**Comment by smuzaffar** [comment]
@iarspider , can you please please take care of (1) in https://github.com/cms-sw/cmsdist/pull/7735#issuecomment-1092576925 ?

**Comment by iarspider** [comment]
@civanch we have tested vecgeom 1.1.20 in master branch (with geant4 10.7.2) in #7775, the failures are present there as well.

**Comment by smuzaffar** [comment]
test parameters:
- full_cmssw = true

**Comment by smuzaffar** [comment]
please test

**Comment by smuzaffar** [comment]
please test

**Comment by smuzaffar** [comment]
please test

**Comment by smuzaffar** [comment]
please test

**Comment by smuzaffar** [comment]
@civanch , this time I build vecgeom with buildin veccore but it still fails with the same error

**Comment by civanch** [comment]
@smuzaffar , may we have some bad interference with dd4hep build? The recent does not explicitly depend on vecgeom, only on geant4.

**Comment by smuzaffar** [comment]
test parameters:
- full_cmssw = true

**Comment by smuzaffar** [comment]
please test

**Comment by smuzaffar** [comment]
let merge this for G4VECGEOM Ib and build an IB to debug it. We can revert it back if needed

**Comment by smuzaffar** [comment]
@civanch , if I build geant4 and vecgeom in `Debug` mode then workflow `1.0` does not fail. In `Release` mode it fails and `valgrind` shows 
```
/afs/cern.ch/user/m/muzaffar/public/geant4-valgrind.log
==23410== Invalid read of size 8
==23410==    at 0x3FD779BA: vecgeom::cxx::CommonUnplacedVolumeImplHelper<vecgeom::cxx::PolyhedronImplementation<(EInnerRadii)0, (EPhiCutout)0>, vecgeom::cxx::VUnplacedVolume>::SafetyToIn(vecgeom::cxx::Vector3D<double> const&) const (in /cvmfs/cms-ib.cern.ch/nweek-02731/slc7_amd64_gcc10/cms/cmssw/CMSSW_12_4_G4VECGEOM_X_2022-05-04-1100/biglib/slc7_amd64_gcc10/pluginSimulation.so)
==23410==    by 0x3FD7B2A6: G4UAdapter<vecgeom::cxx::UnplacedPolyhedron>::DistanceToIn(CLHEP::Hep3Vector const&) const (in /cvmfs/cms-ib.cern.ch/nweek-02731/slc7_amd64_gcc10/cms/cmssw/CMSSW_12_4_G4VECGEOM_X_2022-05-04-1100/biglib/slc7_amd64_gcc10/pluginSimulation.so)
==23410==    by 0x3FE45C48: G4VoxelNavigation::ComputeStep(CLHEP::Hep3Vector const&, CLHEP::Hep3Vector const&, double, double&, G4NavigationHistory&, bool&, CLHEP::Hep3Vector&, bool&, bool&, G4VPhysicalVolume**, int&) (in /cvmfs/cms-ib.cern.ch/nweek-02731/slc7_amd64_gcc10/cms/cmssw/CMSSW_12_4_G4VECGEOM_X_2022-05-04-1100/biglib/slc7_amd64_gcc10/pluginSimulation.so)
==23410==    by 0x3FA5A51D: G4Navigator::ComputeStep(CLHEP::Hep3Vector const&, CLHEP::Hep3Vector const&, double, double&) (in /cvmfs/cms-ib.cern.ch/nweek-02731/slc7_amd64_gcc10/cms/cmssw/CMSSW_12_4_G4VECGEOM_X_2022-05-04-1100/biglib/slc7_amd64_gcc10/pluginSimulation.so)
==23410==    by 0x3FC6F953: G4Transportation::AlongStepGetPhysicalInteractionLength(G4Track const&, double, double, double&, G4GPILSelection*) (in /cvmfs/cms-ib.cern.ch/nweek-02731/slc7_amd64_gcc10/cms/cmssw/CMSSW_12_4_G4VECGEOM_X_2022-05-04-1100/biglib/slc7_amd64_gcc10/pluginSimulation.so)
==23410==    by 0x401BD7E3: G4SteppingManager::DefinePhysicalStepLength() (in /cvmfs/cms-ib.cern.ch/nweek-02731/slc7_amd64_gcc10/cms/cmssw/CMSSW_12_4_G4VECGEOM_X_2022-05-04-1100/biglib/slc7_amd64_gcc10/pluginSimulation.so)
==23410==    by 0x401BC10F: G4SteppingManager::Stepping() (in /cvmfs/cms-ib.cern.ch/nweek-02731/slc7_amd64_gcc10/cms/cmssw/CMSSW_12_4_G4VECGEOM_X_2022-05-04-1100/biglib/slc7_amd64_gcc10/pluginSimulation.so)
==23410==    by 0x3FC66F87: G4TrackingManager::ProcessOneTrack(G4Track*) (in /cvmfs/cms-ib.cern.ch/nweek-02731/slc7_amd64_gcc10/cms/cmssw/CMSSW_12_4_G4VECGEOM_X_2022-05-04-1100/biglib/slc7_amd64_gcc10/pluginSimulation.so)
==23410==    by 0x3F942408: G4EventManager::DoProcessing(G4Event*) (in /cvmfs/cms-ib.cern.ch/nweek-02731/slc7_amd64_gcc10/cms/cmssw/CMSSW_12_4_G4VECGEOM_X_2022-05-04-1100/biglib/slc7_amd64_gcc10/pluginSimulation.so)
==23410==    by 0x3F5C946D: RunManagerMTWorker::produce(edm::Event const&, edm::EventSetup const&, RunManagerMT&) (in /cvmfs/cms-ib.cern.ch/nweek-02731/slc7_amd64_gcc10/cms/cmssw/CMSSW_12_4_G4VECGEOM_X_2022-05-04-1100/biglib/slc7_amd64_gcc10/pluginSimulation.so)
==23410==    by 0x3F5DF634: omt::ThreadHandoff::Functor<OscarMTProducer::produce(edm::Event&, edm::EventSetup const&)::{lambda()#1}>::execute() (in /cvmfs/cms-ib.cern.ch/nweek-02731/slc7_amd64_gcc10/cms/cmssw/CMSSW_12_4_G4VECGEOM_X_2022-05-04-1100/biglib/slc7_amd64_gcc10/pluginSimulation.so)
==23410==    by 0x3F5D52B9: omt::ThreadHandoff::threadLoop(void*) (in /cvmfs/cms-ib.cern.ch/nweek-02731/slc7_amd64_gcc10/cms/cmssw/CMSSW_12_4_G4VECGEOM_X_2022-05-04-1100/biglib/slc7_amd64_gcc10/pluginSimulation.so)
==23410==    by 0x7085EA4: start_thread (in /usr/lib64/libpthread-2.17.so)
==23410==    by 0x7398B0C: clone (in /usr/lib64/libc-2.17.so)
==23410==  Address 0x58 is not stack'd, malloc'd or (recently) free'd
==23410==

```

**Comment by civanch** [comment]
@smuzaffar , I am just debugging CMSSW_12_4_G4VECGEOM_X_2022-05-04-1100. The crash happens on a boundary of a complex volume with the trace exactly as you show. It is not first track but unlucky track. So, it is VecGeom bug and I just make a ticket to VecGeom:
https://sft.its.cern.ch/jira/browse/VECGEOM-600?jql=project%20%3D%20VECGEOM%20AND%20resolution%20%3D%20Unresolved%20ORDER%20BY%20priority%20DESC%2C%20updated%20DESC 

In discussion with Gabriele Cosmo he suggested to switch from 11.0.1 to 11.0ref04 - April reference version of G4 + new VecGeom 1.2.0. The first includes improvements of handling of complex volumes inside Geant4. The second means also some change in cmake options.

**Comment by smuzaffar** [comment]
thanks @civanch . I guess G4 11.0ref04 is not publicly available, can you please provide the tar file for it?

**Comment by civanch** [comment]
@smuzaffar , there the link: https://cernbox.cern.ch/index.php/s/Lsw38cL4NnJPcWV 
VecGeom GitLab: https://gitlab.cern.ch/VecGeom/VecGeom/-/commit/14d3a5c3d851794da4ca48d7659980dbf73384d4

**Comment by smuzaffar** [comment]
thanks, I am testing these now. By the way do we still need https://github.com/cms-externals/geant4/compare/271d2ffb2bd0a2aa26c4d15bc5e99e50f49cd232...cms/v11.0.1 changes for 11.0Ref04?

**Comment by civanch** [comment]
No, they are merged to the development version.

**Comment by bsunanda** [comment]
Hi Shahzad

I have a tar file on /afs/cern.ch/work/b/banerjee/public/geant4.11.0.r04.tgz Please make use of it. Regards

Sunanda
________________________________

**Pull request merged**